### PR TITLE
Add incremental/full build log info

### PR DESCRIPTION
**Pull request by LucasDvp**
- Change `StreamWriter` into `StreamWriter(FileStream)` to support *read/write sharing* mode;
- Add incremental/full write log code-logical;

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs [hunk 1]
```diff
@@ -592,23 +592,34 @@ private static DependencyGraph ConstructDependencyGraphFromLast(DependencyGraph
 
         private static IncrementalStatus GetBuildInfoIncrementalStatus(BuildInfo cb, BuildInfo lb)
         {
+            var details = String.Empty;
```
Comment (herohua) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs]: >String.Empty [](start = 26, length = 12)

null

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs [hunk 2]
```diff
@@ -592,23 +592,34 @@ private static DependencyGraph ConstructDependencyGraphFromLast(DependencyGraph
...
-            if (cb.CommitFromSHA != lb.CommitToSHA)
+            else
             {
-                return new IncrementalStatus { CanIncremental = false, Details = $"Cannot build incrementally because commit SHA doesn't match. Last build commit: {lb.CommitToSHA}. Current build commit base: {cb.CommitFromSHA}." };
+                canIncremental = true;
             }
-            return new IncrementalStatus { CanIncremental = true };
+
+            // Log
+            Logger.LogInfo($"Build strategy", code: canIncremental ? WarningCodes.Build.IsIncrementalBuild : WarningCodes.Build.IsFullBuild);
```
Comment (herohua) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs]: >LogInfo [](start = 19, length = 7)

print strategy in message itself

File: src/Microsoft.DocAsCode.Common/Loggers/ReportLogListener.cs [hunk 3]
```diff
@@ -23,7 +23,8 @@ public ReportLogListener(string reportPath, string repoRoot, string root)
             {
                 Directory.CreateDirectory(dir);
             }
-            _writer = new StreamWriter(reportPath, true);
+            var fs = new FileStream(reportPath, FileMode.OpenOrCreate, FileAccess.Write, FileShare.Read);
```
Comment (herohua) [inline src/Microsoft.DocAsCode.Common/Loggers/ReportLogListener.cs]: >OpenOrCreate [](start = 57, length = 12)

Append

File: src/Microsoft.DocAsCode.Common/Loggers/ReportLogListener.cs [hunk 4]
```diff
@@ -23,7 +23,8 @@ public ReportLogListener(string reportPath, string repoRoot, string root)
...
+            _writer = new StreamWriter(fs);
```
Comment (herohua) [inline src/Microsoft.DocAsCode.Common/Loggers/ReportLogListener.cs]: >StreamWriter [](start = 26, length = 12)

do we really need to change this line? 

according to reference source, `new StreamWriter(reportPath, true)` == `new FileStream(reportPath, FileMode.Append, FileAccess.Write, FileShare.Read)`
[https://referencesource.microsoft.com/#mscorlib/system/io/streamwriter.cs,226](https://referencesource.microsoft.com/#mscorlib/system/io/streamwriter.cs,226)

File: src/Microsoft.DocAsCode.Common/Loggers/WarningCodes.cs [hunk 5]
```diff
@@ -30,6 +30,8 @@ public static class Build
             public const string UidNotFound = "UidNotFound";
             public const string UnknownContentType = "UnknownContentType";
             public const string UnknownContentTypeForTemplate = "UnknownContentTypeForTemplate";
+            public const string IsFullBuild = "IsFullBuild";
+            public const string IsIncrementalBuild = "IsIncrementalBuild";
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Common/Loggers/WarningCodes.cs]: put these into `InfoCodes.cs` as we are using `LogInfo`

**Comment by superyyrrzz** [comment]
bind github account with ms account in opensource.microsoft.com, or sign the cla

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs (hunk 1 above)
Comment (LucasDvp) [reply src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs]: Solved.

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs (hunk 2 above)
Comment (LucasDvp) [reply src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs]: Added.

File: src/Microsoft.DocAsCode.Common/Loggers/ReportLogListener.cs (hunk 4 above)
Comment (LucasDvp) [reply src/Microsoft.DocAsCode.Common/Loggers/ReportLogListener.cs]: Reversed.

File: src/Microsoft.DocAsCode.Common/Loggers/WarningCodes.cs (hunk 5 above)
Comment (LucasDvp) [reply src/Microsoft.DocAsCode.Common/Loggers/WarningCodes.cs]: Changed

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs (hunk 1 above)
Comment (herohua) [reply src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs]: how about `string details = null`

---
In reply to: [200141346](https://github.com/dotnet/docfx/pull/3002#discussion_r200141346) [](ancestors = 200141346)

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs [hunk 6]
```diff
@@ -592,23 +592,35 @@ private static DependencyGraph ConstructDependencyGraphFromLast(DependencyGraph
...
+            else
             {
-                return new IncrementalStatus { CanIncremental = false, Details = $"Cannot build incrementally because commit SHA doesn't match. Last build commit: {lb.CommitToSHA}. Current build commit base: {cb.CommitFromSHA}." };
+                canIncremental = true;
             }
-            return new IncrementalStatus { CanIncremental = true };
+
+            var buildStrategy = canIncremental ? InfoCodes.Build.IsIncrementalBuild : InfoCodes.Build.IsFullBuild;
+            // Log
+            Logger.LogInfo($"Build strategy: {buildStrategy}", code: buildStrategy);
```
Comment (herohua) [inline src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs]: this comment is meaningless

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs (hunk 2 above)
Comment (herohua) [reply src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs]: also add why

---
In reply to: [200141550](https://github.com/dotnet/docfx/pull/3002#discussion_r200141550) [](ancestors = 200141550)

File: src/Microsoft.DocAsCode.Common/Loggers/InfoCodes.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.DocAsCode.Common
+{
+    public static class InfoCodes
```
Comment (herohua) [inline src/Microsoft.DocAsCode.Common/Loggers/InfoCodes.cs]: >InfoCodes [](start = 24, length = 9)

rename to `LogCode`? and move out of `Loggers` folder?

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs (hunk 1 above)
Comment (LucasDvp) [reply src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs]: This is for the **left-align**.

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs (hunk 6 above)
Comment (LucasDvp) [reply src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs]: Deleted

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs (hunk 2 above)
Comment (LucasDvp) [reply src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs]: You means more `message` information?

File: src/Microsoft.DocAsCode.Common/Loggers/InfoCodes.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Common/Loggers/InfoCodes.cs]: add copyright header, move using into namespace

Comment (LucasDvp) [reply src/Microsoft.DocAsCode.Common/Loggers/InfoCodes.cs]: Done

File: src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs (hunk 1 above)
Comment (LucasDvp) [reply src/Microsoft.DocAsCode.Build.Engine/Incrementals/IncrementalBuildContext.cs]: Done

**Pull request merged**